### PR TITLE
runtime: add txn logs

### PR DESCRIPTION
**Pull request by 0x0ece**
Most of the txn log infrastructure is there:
- Logging is disabled in runtime => no impact on perf
- Logging is enabled in testing
- Logs are serialized in protobuf, expect to be binary-compatible with Agave
- Successful txs have logs (syscalls, runtime, bpf loader all are hooked)
- Store meta.logMessages

TODO
   - program_failure and different types of errors

### CODE REVIEW

File: src/flamenco/runtime/context/fd_exec_instr_ctx.h [hunk 1]
```diff
@@ -20,12 +20,22 @@ struct __attribute__((aligned(8UL))) fd_exec_instr_ctx {
...
-  uint instr_err;
+  uint instr_err;  /* TODO: this is kind of redundant wrt instr_exec */
 
   fd_funk_txn_t * funk_txn;
   fd_acc_mgr_t *  acc_mgr;
   fd_valloc_t     valloc;
 
+  /* Most instructions log the base58 program id multiple times, so it's
+     convenient to compute it once and reuse it. */
+  char program_id_base58[ 45 ];
```
Comment (mjain-jump) [inline src/flamenco/runtime/context/fd_exec_instr_ctx.h:31]: probably can use `FD_BASE58_ENCODED_32_SZ` macro

File: src/flamenco/runtime/context/fd_exec_instr_ctx.h [hunk 2]
```diff
@@ -20,12 +20,22 @@ struct __attribute__((aligned(8UL))) fd_exec_instr_ctx {
...
+  /* Most instructions log the base58 program id multiple times, so it's
+     convenient to compute it once and reuse it. */
+  char program_id_base58[ FD_BASE58_ENCODED_32_SZ ];
+
+  /* Execution results (err)
+     TODO: we store `custom_err` inside txn_ctx, so maybe we should
+     store these inside txn_ctx as well? */
+  int vm_exec;
+  int instr_exec;
+
```
Comment (jumpsiegel) [inline src/flamenco/runtime/context/fd_exec_instr_ctx.h:38]: probably..

File: src/flamenco/log_collector/fd_log_collector.h [hunk 3]
```diff
@@ -0,0 +1,526 @@
...
+  /* Copy msg and update total buf_sz */
+  ulong msg_start = buf_sz + 2 + needs_2b;
+  fd_memcpy( buf + msg_start, msg, msg_sz );
+  log->buf_sz = (ushort)( msg_start + msg_sz );
+}
+
+/* _fd_log_collector_debug prints all logs (internal, don't use directly). */
+static inline void
+_fd_log_collector_debug( fd_log_collector_t const * log );
+
```
Comment (jumpsiegel) [inline src/flamenco/log_collector/fd_log_collector.h:57]: Learning your coding style here... what does prefixing with _ represent?

File: src/flamenco/runtime/program/fd_config_program.c [hunk 4]
```diff
@@ -228,11 +230,7 @@ fd_config_program_execute( fd_exec_instr_ctx_t * ctx ) {
 
   /* https://github.com/solana-labs/solana/blob/v1.17.27/programs/config/src/config_processor.rs#L14
      See DEFAULT_COMPUTE_UNITS */
-
-  do {
-    int err = fd_exec_consume_cus( ctx->txn_ctx, 450UL );
-    if( FD_UNLIKELY( err ) ) return err;
-  } while(0);
+  FD_EXEC_CU_UPDATE( ctx, DEFAULT_COMPUTE_UNITS );
```
Comment (jumpsiegel) [inline src/flamenco/runtime/program/fd_config_program.c:234]: good catch

**Review by jumpsiegel — APPROVED** [review]

File: src/flamenco/runtime/fd_executor.c [hunk 5]
```diff
@@ -1524,7 +1544,7 @@ fd_execute_txn( fd_exec_txn_ctx_t * txn_ctx ) {
         ret = fd_sysvar_instructions_update_current_instr_idx( txn_ctx, i );
         if( ret != FD_ACC_MGR_SUCCESS ) {
           FD_LOG_WARNING(( "sysvar instructions failed to update instruction index" ));
-          return ret;
+          goto ret;
```
Comment (asiegel-jt) [inline src/flamenco/runtime/fd_executor.c:1547]: use a variable name as a label is very confusing

File: src/flamenco/runtime/context/fd_exec_instr_ctx.h (hunk 2 above)
Comment (0x0ece) [reply src/flamenco/runtime/context/fd_exec_instr_ctx.h:38]: I'll work with Ishan to handle this

File: src/flamenco/log_collector/fd_log_collector.h (hunk 3 above)
Comment (0x0ece) [reply src/flamenco/log_collector/fd_log_collector.h:57]: Usually private methods, but I see the confusion - renamed it to private so it's more clean

File: src/flamenco/runtime/fd_executor.c (hunk 5 above)
Comment (0x0ece) [reply src/flamenco/runtime/fd_executor.c]: good point: renamed as `exit`

**Review by asiegel-jt — DISMISSED** [review]

**Review by jumpsiegel — APPROVED** [review]

File: src/flamenco/vm/syscall/fd_vm_syscall_util.c [hunk 6]
```diff
@@ -17,38 +18,58 @@ fd_vm_syscall_abort( FD_PARAM_UNUSED void *  _vm,
...
-  return FD_VM_ERR_ABORT;
-}
+   As of v0.2.6, the only two usages are in syscall panic and syscall log. */
+#define FD_TRANSLATE_STRING( vm, vaddr, msg_sz ) (__extension__({          \
+    char const * msg = FD_VM_MEM_SLICE_HADDR_LD( vm, vaddr, 1UL, msg_sz ); \
+    if( FD_UNLIKELY( !fd_utf8_verify( msg, msg_sz ) ) ) {                  \
+      return FD_VM_ERR_INVAL; /* Err(SyscallError::InvalidString(...)) */  \
+    }                                                                      \
+    msg;                                                                   \
+}))
```
Comment (topointon-jump) [inline src/flamenco/vm/syscall/fd_vm_syscall_util.c:38]: Be aware that if msg_sz is zero than msg is an undefined pointer (as described in the documentation of FD_VM_MEM_SLICE_HADDR_LD).

**Review by mjain-jump — DISMISSED** [review]

File: src/flamenco/vm/syscall/fd_vm_syscall_util.c (hunk 6 above)
Comment (0x0ece) [reply src/flamenco/vm/syscall/fd_vm_syscall_util.c:38]: That's ok, `fd_utf8_verify` works with msg_sz==0.

Comment (topointon-jump) [reply src/flamenco/vm/syscall/fd_vm_syscall_util.c:38]:
```rust
  char const * file = FD_TRANSLATE_STRING( vm, file_vaddr, file_sz );

  /* Max msg_sz: 37 - 6 + 20 + 20 + file_sz = 71 + file_sz <= 127 => file_sz <= 56 */
  if( file_sz<=56 ) {
    fd_log_collector_printf_dangerous_max_127( vm->instr_ctx,
      "SBF program Panicked in %s at %lu:%lu", file, line, column );
  }
  ```

In the case where file_sz==0, we pass a potentially undefined pointer to fd_log_collector_printf_dangerous_max_127 - is this safe?

Comment (0x0ece) [reply src/flamenco/vm/syscall/fd_vm_syscall_util.c:38]: Oh right - fixed with "%.*s", run some tests to confirm. We definitely need some test vectors for panic.

File: src/flamenco/log_collector/fd_log_collector_base.h [hunk 7]
```diff
@@ -0,0 +1,37 @@
+#ifndef HEADER_fd_src_flamenco_log_collector_fd_log_collector_base_h
+#define HEADER_fd_src_flamenco_log_collector_fd_log_collector_base_h
+
+#include "../fd_flamenco_base.h"
+
+/* Base definition for fd_log_collector_t.
+   (This is needed to avoid circular dependencies) */
+
+#define FD_LOG_COLLECTOR_MAX   (10000UL)  /* Max bytes of actual log messages before truncate */
```
[the remark this answers is not in the document]
Comment (0x0ece) [reply src/flamenco/log_collector/fd_log_collector_base.h]: Added comments, 10k is from agave. The other is just a rough approximation to stay safe.

**Review by topointon-jump — DISMISSED** [review]

**Review by lheeger-jump — CHANGES_REQUESTED** [review]
. will affect perf, nack until tomorrow at a min

**Comment by asiegel-jt** [comment]
to make this more efficient, the transaction logs need to be allocated directly out of the blockstore workspace so we just have to move pointers at the end. note that space in this workspace can be tight, so don't overallocate buffers too much

**Review by asiegel-jt — APPROVED** [review]

File: src/flamenco/runtime/context/fd_exec_txn_ctx.h [hunk 8]
```diff
@@ -109,6 +110,8 @@ struct __attribute__((aligned(8UL))) fd_exec_txn_ctx {
 
   fd_exec_instr_trace_entry_t instr_trace[FD_MAX_INSTRUCTION_TRACE_LENGTH]; /* Instruction trace */
   ulong                       instr_trace_length;                           /* Number of instructions in the trace */
+
+  fd_log_collector_t          log_collector;             /* Log collector instance */
```
[the remark this answers is not in the document]
Comment (lheeger-jump) [reply src/flamenco/runtime/context/fd_exec_txn_ctx.h:114]: No. Workspace allocation is too slow

File: src/flamenco/log_collector/fd_log_collector_base.h [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+
+  ushort log_sz;   /* The total bytes count of logs inserted, up to
+                      FD_LOG_COLLECTOR_MAX.
+                      This is only used to match Agave's behavior and
+                      truncate logs when necessary. */
+
+  uchar  warn;     /* Whether we truncated or not logs, to match Agave's
+                      behavior. */
+
+  uchar  disabled; /* Whether txn logs are disabled (1) or enabled (0). */
```
Comment (lheeger-jump) [inline src/flamenco/log_collector/fd_log_collector_base.h:32]: who is setting these values. when is `log_sz` set to 0?

**Review by lheeger-jump — CHANGES_REQUESTED** [review]

File: src/flamenco/log_collector/fd_log_collector_base.h (hunk 9 above)
Comment (0x0ece) [reply src/flamenco/log_collector/fd_log_collector_base.h:32]: fd_log_collector_init()

**Review by lheeger-jump — APPROVED** [review]

**Pull request merged**